app/waf/src/services/bigiq.service.ts

```diff
@@ -0,0 +1,167 @@
+import {checkAndWait} from '../utils';
+
+const BIGIQ_HOST: string = process.env.BIGIQ_HOST || 'localhost';
+const BIGIQ_PORT: number = Number(process.env.BIGIQ_PORT) || 8888;
+const BIGIQ_USERNAME: string = process.env.BIGIQ_USERNAME || 'admin';
+const BIGIQ_PASSWORD: string = process.env.BIGIQ_PASSWORD || 'admin';
+const BIGIQ_POOL: string = process.env.BIGIQ_POOL || 'unknown';
+
+type BigIqToken = {
+  token: string;
+  exp: number;
+};
+
+type LoginResponse = {
+  token: BigIqToken;
+};
+
+type RevokeResponse = {
+  id: string;
+};
+
+type QueryResponse = {
+  status: string;
+  errorMessage: string;
+};
+
+export interface BigIqService {
+  login(
+    host: string,
+    port: number,
+    username: string,
+    password: string,
+  ): Promise<LoginResponse>;
+
+  revoke(
+    host: string,
+    port: number,
+    token: string,
+    body: object,
+  ): Promise<RevokeResponse>;
+
+  query(
+    host: string,
+    port: number,
+    token: string,
+    task: string,
+  ): Promise<QueryResponse>;
+}
+
+export class BigIqServiceProvider implements Provider<BigIqService> {
+  constructor(
+    @inject('datasources.bigiq')
+    protected dataSource: BigIqDataSource = new BigIqDataSource(),
+  ) {}
+
+  value(): Promise<BigIqService> {
+    return getService(this.dataSource);
+  }
+}
+
+export class BigIqManager {
+  private service: BigIqService;
+  private logger: Logger;
+  private static token: BigIqToken;
+
+  constructor(svc: BigIqService, reqId = 'Unknown') {
+    this.service = svc;
+    this.logger = factory.getLogger(reqId + ': services.BigIqManager');
+  }
+
+  async login(): Promise<BigIqToken> {
+    let current = new Date().getTime() / 1000;
+    // Login BIG-IQ, if no token or token will expire in 30 seconds
+    if (!BigIqManager.token || current > BigIqManager.token.exp - 30) {
+      let resp = await this.service.login(
+        BIGIQ_HOST,
+        BIGIQ_PORT,
+        BIGIQ_USERNAME,
+        BIGIQ_PASSWORD,
+      );
+      BigIqManager.token = resp.token;
+      this.logger.debug(`BIG-IQ token will expire at ${resp.token.exp}`);
+    }
+
+    return BigIqManager.token;
+  }
+
+  async revokeLicense(
+    address: string,
+    mac: string = '',
+    pool: string = BIGIQ_POOL,
+  ): Promise<void> {
+    let body = {
+      licensePoolName: pool,
+      command: 'revoke',
+      address: address,
+    };
+
+    if (mac) {
+      // Unreachable license
+      Object.assign(body, {
+        assignmentType: 'UNREACHABLE',
+        macAddress: mac.toUpperCase(),
+      });
+    } else {
+      // Unmanaged license
+      Object.assign(body, {
+        user: 'wrong',
+        password: 'wrong',
+      });
+    }
+
+    return this.revoke(body);
+  }
+
+  async revoke(body: object): Promise<void> {
+    let token = await this.login();
+    let revokeResp = await this.service.revoke(
+      BIGIQ_HOST,
+      BIGIQ_PORT,
+      token.token,
+      body,
+    );
+
+    let taskFinished = async (taskId: string): Promise<boolean> => {
```

Comment:
The parameter `taskId: string` is useless here, since you can capture `revokeResp.id` within the inline function. 

By the way, function `checkAndWait`'s 3rd parameter is useless, and it may lead to potential issues.

Reply:
done